counsel-gtags.el

```diff
@@ -414,70 +348,89 @@ This is for internal use and not for final user."
 			  (`definition '())
-			  (_ (list tagname))))
-	 (global-args (append (reverse options) query-as-list)))
-    (apply #'counsel-gtags--process-lines "global" global-args)))
-
-(defun counsel-gtags--select-file-ivy-parameters (type tagname extra-options auto-select-only-candidate)
-  "Get `counsel-gtags--select-file' the parameters from TYPE to call `ivy-read'."
-  (if (string-empty-p tagname)
-      (message "No candidate tags")
-    (let* ((root (counsel-gtags--default-directory))
-           (encoding buffer-file-coding-system)
-           (default-directory root)
-           (collection (counsel-gtags--collect-candidates
-			type tagname encoding extra-options))
-           (ivy-auto-select-single-candidate t) ;; see issue #7
-           )
-      `("Pattern: " ,collection
-        :action counsel-gtags--find-file))))
-
-(defun counsel-gtags--select-file (type tagname &optional extra-options auto-select-only-candidate)
+			  (_ (shell-quote-argument tagname))))
+	 (global-args (concat options query-as-list)))
+    (counsel-gtags--process-lines "global" global-args)))
+
+(defsubst counsel-gtags--select-file-collection (type tagname extra-options)
+  "Candidated collection for counsel-gtags--select-file."
+  (counsel-gtags--collect-candidates
+   type tagname buffer-file-coding-system extra-options))
+
+(defun counsel-gtags--select-file (type tagname
+					&optional extra-options auto-select-only-candidate)
   "Prompt the user to select a file_path:position according to query.
 
 Use TYPE ∈ '(definition reference symbol) for defining global parameters.
 Use TAGNAME for global query.
 Use AUTO-SELECT-ONLY-CANDIDATE to skip `ivy-read' if have a single candidate.
 Extra command line parameters to global are forwarded through EXTRA-OPTIONS."
-  (let* ((the-ivy-arguments
-	  (counsel-gtags--select-file-ivy-parameters type
-						     tagname
-						     extra-options
-						     auto-select-only-candidate))
-	 (collection (cadr the-ivy-arguments)))
+  (let* ((default-directory (counsel-gtags--default-directory))
+	 (collection (counsel-gtags--select-file-collection type tagname extra-options))
+	 (ivy-auto-select-single-candidate t)
+	 (first (cadr collection)))
     (if (and auto-select-only-candidate (= (length collection) 1))
-        (counsel-gtags--find-file (car collection))
-      ;; else
-      (apply 'ivy-read (plist-put
-			the-ivy-arguments
-			:caller 'counsel-gtags--select-file)))))
+        (counsel-gtags--find-file (car first))
+      (ivy-read "Pattern: "
+		collection
+		:action #'counsel-gtags--find-file
+		:caller 'counsel-gtags--select-file))))
+
+(ivy-set-actions
+ 'counsel-gtags--select-file
+ '(("j" counsel-gtags--find-file-other-window "other window")))
 
 ;;;###autoload
 (defun counsel-gtags-find-definition (tagname)
   "Search for TAGNAME definition in tag database.
 Prompt for TAGNAME if not given."
   (interactive
-   (list (counsel-gtags--read-tag 'definition)))
+   (list (counsel-gtags--read-tag definition)))
```

Comment:
Actually no, We shouldn't pass 'whatever to macros when we expect the macro expands the value.

Otherwise `,(alist-get type counsel-gtags--prompts)` will expand wrongly.

Reply:
I don't if executing macros as function calls is a good idea. This is my personal take, but I'm not stopping this code from being merged.